source/common/tcp_proxy/tcp_proxy.cc

```diff
@@ -440,47 +441,36 @@ Network::FilterStatus Filter::initializeUpstreamConnection() {
   return Network::FilterStatus::StopIteration;
 }
 
-bool Filter::maybeTunnel(const std::string& cluster_name) {
-  if (!config_->tunnelingConfig()) {
-    generic_conn_pool_ =
-        std::make_unique<TcpConnPool>(cluster_name, cluster_manager_, this, *upstream_callbacks_);
-    if (generic_conn_pool_->valid()) {
-      connecting_ = true;
-      connect_attempts_++;
-      generic_conn_pool_->newStream(this);
-      // Because we never return open connections to the pool, this either has a handle waiting on
-      // connection completion, or onPoolFailure has been invoked. Either way, stop iteration.
-      return true;
-    } else {
-      generic_conn_pool_.reset();
-    }
+bool Filter::maybeTunnel(Upstream::ThreadLocalCluster& cluster, const std::string& cluster_name) {
+  GenericConnPoolFactory* factory = nullptr;
+  if (cluster.info()->upstreamConfig().has_value()) {
+    factory = Envoy::Config::Utility::getFactory<GenericConnPoolFactory>(
+        cluster.info()->upstreamConfig().value());
   } else {
-    auto* cluster = cluster_manager_.get(cluster_name);
-    if (!cluster) {
-      return false;
-    }
-    // TODO(snowp): Ideally we should prevent this from being configured, but that's tricky to get
-    // right since whether a cluster is invalid depends on both the tcp_proxy config + cluster
-    // config.
-    if ((cluster->info()->features() & Upstream::ClusterInfo::Features::HTTP2) == 0) {
-      ENVOY_LOG(error, "Attempted to tunnel over HTTP/1.1, this is not supported. Set "
-                       "http2_protocol_options on the cluster.");
-      return false;
-    }
-
-    generic_conn_pool_ = std::make_unique<HttpConnPool>(cluster_name, cluster_manager_, this,
-                                                        config_->tunnelingConfig()->hostname(),
-                                                        *upstream_callbacks_);
-    if (generic_conn_pool_->valid()) {
-      generic_conn_pool_->newStream(this);
-      return true;
-    } else {
-      generic_conn_pool_.reset();
-    }
+    factory = Envoy::Config::Utility::getFactoryByName<GenericConnPoolFactory>(
+        "envoy.filters.connection_pools.tcp.generic");
+  }
+  if (!factory) {
+    return false;
   }
 
+  absl::optional<std::string> hostname =
```

Comment:
This does a string copy, which means two string copies here since hostname is passed by value to `createGenericConnPool`. Might not matter in this context

A higher level comment: should we do the factory lookup + partially populate the arguments during filter factory creation? That way we avoid factory lookup and string copies etc in the data path